### PR TITLE
Dpd charged

### PR DESCRIPTION
**Pull request by Eddy-Barraud**
**Summary**

DPD simulations with charged interactions, using coul/slater/long pair style, requires the use of hybrid/overlay pair style. This option is forcing host neighboring (#2621 ) which is a bottleneck preventing any significant GPU acceleration of simulations. 
As I really need to simulate big systems, and because of my work environment, I suggest the creation of a new fix including DPD+Slater and allowing device neighboring.

The new fix is called dpd/charged and I successfully ran CPU calculations. Although, there is a divergence in energies after 500 steps which I don't really explain.

The GPU acceleration code (dpd/charged/gpu), is compiling but fail at simulation run time with the following error code:
```shell
Caught signal 11 (Segmentation fault: address not mapped to object at address 0x4)
==== backtrace (tid: 807733) ====
 0 0x000000000003e6f0 __GI___sigaction()  :0
 1 0x000000000147b1b1 LAMMPS_AL::DPDCharged<float, double>::loop()  /work/barraude/LAMMPS/lammps-custom/lib/gpu/lal_dpd_charged.cpp:180
 2 0x000000000145a7f3 LAMMPS_AL::BaseDPD<float, double>::compute()  /work/barraude/LAMMPS/lammps-custom/lib/gpu/lal_base_dpd.cpp:317
 3 0x000000000140d501 dpd_charged_gpu_compute_n()  /work/barraude/LAMMPS/lammps-custom/lib/gpu/lal_dpd_charged_ext.cpp:109
 4 0x0000000001398757 LAMMPS_NS::PairDPDChargedGPU::compute()  /work/barraude/LAMMPS/lammps-custom/src/GPU/pair_dpd_charged_gpu.cpp:258
 5 0x000000000061174a LAMMPS_NS::Verlet::setup()  /work/barraude/LAMMPS/lammps-custom/src/verlet.cpp:140
 6 0x000000000059d8cd LAMMPS_NS::Run::command()  /work/barraude/LAMMPS/lammps-custom/src/run.cpp:171
 7 0x000000000048b9ff LAMMPS_NS::Input::execute_command()  /work/barraude/LAMMPS/lammps-custom/src/input.cpp:868
 8 0x000000000048c14e LAMMPS_NS::Input::file()  /work/barraude/LAMMPS/lammps-custom/src/input.cpp:313
 9 0x000000000047bf1d main()  /work/barraude/LAMMPS/lammps-custom/src/main.cpp:77
10 0x0000000000029590 __libc_start_call_main()  ???:0
11 0x0000000000029640 __libc_start_main_alias_2()  :0
12 0x00000000004802f5 _start()  ???:0
=================================
``` 
It seems to be related to the charge of atoms.
I tried to implement the passing of extra atom information like in the mdpd pair style, as suggested by Trung.

I am still trying to resolve this error and seek for any help.

**Related Issue(s)**

#2621 

**Author(s)**

Eddy BARRAUD <eddy.barraud@outlook.fr>
Trung Nguyen <ndactrung@gmail.com>

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

no

**Implementation Notes**

I merged dpd code with coul/slater/long code.

The coeff function defines the global cutoff of a pair but also a DPD cutoff and a Slater cutoff which is equal to zero if the boolean of pair_coeff is set to no.

The init_one function is inspired by pair_lj_cut_coul_cut with the definition of a global cutoff which is the maximum between DPD and Slater cutoff.

The new pair_style has those parameters:
pair_style dpd/charged/gpu {T} {cut_dpd} {seed} {Slater Lambda} {cut_slater}

The pair_coeff are defined with an optional parameter (Boolean) for the charged-charged species. It should be set to yes to activate the Slater calculation. This allows to set the optimal cutoff of neighboring for each pair of bead types.
pair_coeff    i   j   {a_0}  {gamma} {Bool Slater}

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [x] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

Test case with the creation of a random box (box.create.lmps), then different input files for hybrid/overlay, dpd/charged, dpd/charged/gpu: 
[test.case.tar.gz](https://github.com/user-attachments/files/15661271/test.case.tar.gz)

### CODE REVIEW

**Comment by ndtrung81** [comment]
@Eddy-Barraud you need to allocate the atom extra  with `extra_fields` (similar to what is done with lal_mdpd.cpp). The changes are below
```
diff --git a/lib/gpu/lal_dpd_charged.cpp b/lib/gpu/lal_dpd_charged.cpp
index d8dff0a6e9..9673c4fc86 100644
--- a/lib/gpu/lal_dpd_charged.cpp
+++ b/lib/gpu/lal_dpd_charged.cpp
@@ -72,8 +72,10 @@ int DPDChargedT::init(const int ntypes,
   #endif
 
   int success;
+  int extra_fields = 4; // round up to accomodate quadruples of numtyp values
+                        // q
   success=this->init_atomic(nlocal,nall,max_nbors,maxspecial,cell_size,
-                            gpu_split,_screen,dpd_charged,"k_dpd_charged",onetype);
+                            gpu_split,_screen,dpd_charged,"k_dpd_charged",onetype,extra_fields);
   if (success!=0)
     return success;
```

**Comment by Eddy-Barraud** [comment]
Thank you @ndtrung81 that solved the seg fault.

Now I have a new problem, the coulombic energies are not reported in the thermo output:
```
   Step          Temp          Press          Volume         E_vdwl         E_coul         E_long         PotEng         KinEng         Fnorm           Fmax     
         0   1              49.455608      125            4477.8549      0             -21.425463      4456.4294      561            579.25252      89.686232    
       100   0.89616095     46.821903      125            4683.0555      0             -28.459031      4654.5965      502.74629      621.12235      62.022875    
       200   0.87776794     47.820369      125            4686.7137      0             -26.883872      4659.8298      492.42781      599.46003      66.111008    
       300   1.04735        48.118192      125            4672.5522      0             -26.766901      4645.7853      587.56333      622.2823       64.042822    
       400   1.0541613      47.830886      125            4693.1371      0             -26.029017      4667.1081      591.38448      629.43415      62.23897     
       500   1.0851503      48.397028      125            4725.0347      0             -24.684696      4700.35        608.76933      652.24365      69.726272   

``` 

Maybe it's a problem with the `store_answers_q(f,energy,e_coul, ...)` in lal_dpd_charged.cu ?

**Comment by ndtrung81** [comment]
@Eddy-Barraud for debugging purposes, in `lal_base_dpd.cpp`, `init_atomic()` can you change:
```
bool charge = false;
```
to
```
bool charge = true;
```
and rebuild? If this change gives ecoul correctly, I will modify `BaseDPD::init_atomic()` to support atom charges whenever needed.

**Comment by Eddy-Barraud** [comment]
@ndtrung81 I tried and its working now:

```
   Step          Temp          Press          Volume         E_vdwl         E_coul         E_long         PotEng         KinEng         Fnorm           Fmax     
         0   1              70.484366      125            4477.8549     -5.2076243     -21.425463      4451.2218      561            579.25252      89.686232    
       100   0.89616127     67.671597      125            4683.0556     -14.432382     -28.459035      4640.1642      502.74647      621.12201      62.022652    
       200   0.87773024     69.46121       125            4686.7234     -10.204273     -26.883921      4649.6352      492.40666      599.47115      66.108452    
       300   1.0508219      68.964481      125            4671.2854     -10.879881     -26.732233      4633.6733      589.5111       623.97619      65.91799     
       400   1.0088337      69.323057      125            4700.3969     -10.16237      -26.696173      4663.5384      565.95569      652.19092      73.454781    
       500   1.0001401      68.28031       125            4695.2322     -8.6328785     -27.115572      4659.4837      561.07861      655.44497      61.112967
``` 

I still have some energy differencies but now it's really close 😁

**Comment by ndtrung81** [comment]
@Eddy-Barraud cool, looks promising. The random number generator on the dpd  on the GPU is not the same as that on the CPU version, consequently even when you built with double precision the numbers (forces) are not the same and getting diverged rather quickly over timesteps, hence energy diffs.

**Comment by Eddy-Barraud** [comment]
[test.2.beads.tar.gz](https://github.com/user-attachments/files/15759885/test.2.beads.tar.gz)

I tested the energies for 2 charged beads without random numbers `( Gamma = 0 ==> sigma = 0 )` and the energies are the same up to 1E-07 for GPU vs CPU

GPU:
```
------------ Step              0 ----- CPU =            0 (sec) -------------
Step     =              0 Temp     =                    0 Press    =   0.0475193696294839 
Volume   =                  125 E_vdwl   =                 9.75 E_coul   =   -0.030399426817894 
E_long   =    -2.06122451345904 PotEng   =     7.65837605972307 KinEng   =                    0 
Fnorm    =      50.427336482002 Fmax     =     35.6575115835994
``` 

CPU:
```
------------ Step              0 ----- CPU =            0 (sec) -------------
Step     =              0 Temp     =                    0 Press    =   0.0475193697085465 
Volume   =                  125 E_vdwl   =                 9.75 E_coul   =  -0.0303994557372416 
E_long   =    -2.06122451345904 PotEng   =     7.65837603080372 KinEng   =                    0 
Fnorm    =     50.4273365658606 Fmax     =     35.6575116428964
``` 

@ndtrung81 should I open this PR or we wait for the new BaseDPD::init_atomic() ?

**Comment by ndtrung81** [comment]
@Eddy-Barraud can you try the following patch?
```
diff --git a/lib/gpu/lal_base_dpd.cpp b/lib/gpu/lal_base_dpd.cpp
index 0ddd24d21e..dfeaf24ce9 100644
--- a/lib/gpu/lal_base_dpd.cpp
+++ b/lib/gpu/lal_base_dpd.cpp
@@ -57,7 +57,7 @@ int BaseDPDT::init_atomic(const int nlocal, const int nall,
                           const double cell_size, const double gpu_split,
                           FILE *_screen, const void *pair_program,
                           const char *k_name, const int onetype,
-                          const int extra_fields) {
+                          const int extra_fields, bool need_charges) {
   screen=_screen;
 
   int gpu_nbor=0;
@@ -73,7 +73,7 @@ int BaseDPDT::init_atomic(const int nlocal, const int nall,
 
   _threads_per_atom=device->threads_per_atom();
 
-  bool charge = false;
+  bool charge = need_charges;
   bool rot = false;
   bool vel = true;
   _extra_fields = extra_fields;
diff --git a/lib/gpu/lal_base_dpd.h b/lib/gpu/lal_base_dpd.h
index 64ec725d95..658147419d 100644
--- a/lib/gpu/lal_base_dpd.h
+++ b/lib/gpu/lal_base_dpd.h
@@ -53,7 +53,8 @@ class BaseDPD {
                   const int maxspecial, const double cell_size,
                   const double gpu_split, FILE *screen,
                   const void *pair_program, const char *k_name,
-                  const int onetype=0, const int extra_fields=0);
+                  const int onetype=0, const int extra_fields=0,
+                  bool need_charges=false);
 
   /// Estimate the overhead for GPU context changes and CPU driver
   void estimate_gpu_overhead();
```
[base_dpd.diff.txt](https://github.com/user-attachments/files/15776677/base_dpd.diff.txt)

Please rename the files lal_dpd_charged* to lal_dpd_coul_slater_long* (and their CPU pair style counterpart) to be consistent with the sub styles implemented here. Maybe there will be other pair styles that combine dpd with other coul* styles in the future.

Because you also add the CPU version of the combined pair style, you will need to add a doc page for it as well. After these, I think the PR will be ready for review.

Thanks very much for the nice work!

**Comment by ndtrung81** [comment]
.. and some change to `lal_dpd_charged.cpp`:
```
diff --git a/lib/gpu/lal_dpd_charged.cpp b/lib/gpu/lal_dpd_charged.cpp
index 9673c4fc86..a0db324664 100644
--- a/lib/gpu/lal_dpd_charged.cpp
+++ b/lib/gpu/lal_dpd_charged.cpp
@@ -74,8 +74,9 @@ int DPDChargedT::init(const int ntypes,
   int success;
   int extra_fields = 4; // round up to accomodate quadruples of numtyp values
                         // q
+  bool need_charges = true;
   success=this->init_atomic(nlocal,nall,max_nbors,maxspecial,cell_size,
-                            gpu_split,_screen,dpd_charged,"k_dpd_charged",onetype,extra_fields);
+                            gpu_split,_screen,dpd_charged,"k_dpd_charged",onetype,extra_fields,need_charges);
   if (success!=0)
     return success;
```

**Comment by Eddy-Barraud** [comment]
It's working, thank you.
I'll change the name of the fix and write a doc page

**Comment by Eddy-Barraud** [comment]
I didn't try the OPENCL compilation, and I am not sure to understand the code below, even though i changed the name of files.

lal_dpd_coul_slater_long.cpp:
```
#if defined(USE_OPENCL)
#include "dpd_coul_slater_long_cl.h"
#elif defined(USE_CUDART)
const char *dpd_charged=0;
#else
#include "dpd_coul_slater_long_cubin.h"
#endif
``` 
Is it generating the files dpd_coul_slater_long_cl.h and dpd_coul_slater_long_cubin.h if needed during compilation?

I generated a doc, but don't know how to preview it.

**Comment by ndtrung81** [comment]
@Eddy-Barraud For the doc page, you can do `make html`  with the traditional build, or with `-DBUILD_DOC=on` with cmake build. There are instructions at

https://docs.lammps.org/Build_manual.html

also note instructions for developers and the prerequisites depending on your machine. The generated .html output will be under build/doc/html.

Right, the _cl.h and _cubin.h are generated during the build depending on whether you are using OpenCL or CUDA API. For a traditional build, you can see `lib/gpu/Opencl.makefile` to see how the _cl.h files are generated, which converts the .cu file and the included headers into a string. The _cubin.h files are generated with bin2c, as shown in `Nvidia.makefile`. For CMake bulid, you can see `cmake/Modules/Packages/GPU.cmake` for more details, which essentially invoke the corresponding commands.  

```
#if defined(USE_OPENCL)
#include "dpd_coul_slater_long_cl.h"
#elif defined(USE_CUDART)
const char *dpd_coul_slater_long=0;
#else
#include "dpd_coul_slater_long_cubin.h"
#endif
```

**Comment by akohlmey** [comment]
@Eddy-Barraud Please go to the src folder, type "make fix-whitespace" and commit/push the resulting changes. Thanks.

**Comment by Eddy-Barraud** [comment]
> @Eddy-Barraud For the doc page, you can do `make html` with the traditional build, or with `-DBUILD_DOC=on` with cmake build. There are instructions at
> 
> https://docs.lammps.org/Build_manual.html
> [...]

@ndtrung81 Thank you for the answer I successfully build the doc and it seems ok, I wrote the coulombic force with Slater form instead of the potential to sum it in the DPD equation.

@akohlmey I did the fix. I have a quick question for you, will LAMMPS support cuFFT as the heFFTE backend ? I saw it's possible in [this github page](https://github.com/af-ayala/heffte/blob/master/doxygen/installation.md) maybe it could improve kspace calculation time?

**Comment by akohlmey** [comment]
> I have a quick question for you, will LAMMPS support cuFFT as the heFFTE backend ?

No.

**Comment by Eddy-Barraud** [comment]
> > I have a quick question for you, will LAMMPS support cuFFT as the heFFTE backend ?
> 
> No.

Right. Tell me when I should put the PR ready for review.

**Comment by akohlmey** [comment]
> Right. Tell me when I should put the PR ready for review.

You can do so now. There will have to be some changes to integrate your added documentation fully into the manual.
Also, there needs to be an update to the installation settings, since your pair style requires KSPACE while the other DPD styles in DPD-BASIC don't. This will have to be done for both the CMake and GNU make build process. In essence, your style must not be compiled and included unless the KSPACE package is also included.
Finally, there are some updates need to src/.gitignore and .github/CODEOWNERS.

I can look into that, since it is usually far easier for me to do those changes myself than to explain them and find and correct any oversights by others.

**Comment by Eddy-Barraud** [comment]
> > Right. Tell me when I should put the PR ready for review.
> 
> You can do so now. There will have to be some changes to integrate your added documentation fully into the manual. Also, there needs to be an update to the installation settings, since your pair style requires KSPACE while the other DPD styles in DPD-BASIC don't. This will have to be done for both the CMake and GNU make build process. In essence, your style must not be compiled and included unless the KSPACE package is also included. Finally, there are some updates need to src/.gitignore and .github/CODEOWNERS.
> 
> I can look into that, since it is usually far easier for me to do those changes myself than to explain them and find and correct any oversights by others.

Great, I let you do it then. 
Thank you to have explained it, I now have a complete picture of the procedure to create a new pair_style 😁.

**Comment by akohlmey** [comment]
@Eddy-Barraud It would be great if you could construct an example for your (CPU) pair style as a folder in `examples/PACKAGES/dpd-basic` similar to the existing ones with a one-line explanation in the README.

**Comment by ndtrung81** [comment]
@Eddy-Barraud could you also correct the contributing author to be you in the .h and .cpp files under lib/gpu and src/GPU? same for the CPU version.

**Comment by Eddy-Barraud** [comment]
> @Eddy-Barraud could you also correct the contributing author to be you in the .h and .cpp files under lib/gpu and src/GPU? same for the CPU version.

I am doing it. Should I let only your name in the .cu files? For the other files, as you wish, but I would have not do it without your help, thanks again.

**Comment by Eddy-Barraud** [comment]
> @Eddy-Barraud It would be great if you could construct an example for your (CPU) pair style as a folder in `examples/PACKAGES/dpd-basic` similar to the existing ones with a one-line explanation in the README.

@akohlmey I wrote i little input script, it runs correctly. Am I supposed to include the log, or it will be automatically generated?

**Comment by akohlmey** [comment]
> @akohlmey I wrote i little input script, it runs correctly. Am I supposed to include the log, or it will be automatically generated?

It is not automatic, but let me generate them so they are consistent with the rest.

File: lib/gpu/lal_dpd_coul_slater_long.cu [hunk 1]
```diff
@@ -3,7 +3,7 @@
 //                             -------------------
 //                           Trung Dac Nguyen (ORNL)
```
Comment (ndtrung81) [inline lib/gpu/lal_dpd_coul_slater_long.cu]: @Eddy-Barraud please add your name ahead of mine, and replace ORNL with U Chicago.

**Comment by akohlmey** [comment]
> @akohlmey I wrote i little input script, it runs correctly. Am I supposed to include the log, or it will be automatically generated?

I see the note in the README file (BTW, this was missing a newline at the last line). but I don't see the example.

**Comment by Eddy-Barraud** [comment]
> > @akohlmey I wrote i little input script, it runs correctly. Am I supposed to include the log, or it will be automatically generated?
> 
> I see the note in the README file (BTW, this was missing a newline at the last line). but I don't see the example.

My bad, i correctly committed this time.

File: lib/gpu/lal_dpd_coul_slater_long.cu [hunk 2]
```diff
@@ -0,0 +1,523 @@
...
+//                           Trung Dac Nguyen (U Chicago)
+//
+//  Device code for acceleration of the dpd/coul/slater/long pair style
+//
+// __________________________________________________________________________
+//    This file is part of the LAMMPS Accelerator Library (LAMMPS_AL)
+// __________________________________________________________________________
+//
+//    begin                : Jan 15, 2014
+//    email                : nguyentd@ornl.gov
```
Comment (ndtrung81) [inline lib/gpu/lal_dpd_coul_slater_long.cu]: @Eddy-Barraud please use your email and fix the begin date

**Comment by ndtrung81** [comment]
@Eddy-Barraud could you please rename the classes to be consistent with the source files, `PairDPDCharged` to `PairDPDCoulSlaterLong` and their GPU counterpart? Similarly, the class in the GPU library would be `DPDCoulSlaterLong`, instead of `DPDCharged`. Thanks!

**Comment by Eddy-Barraud** [comment]
> @Eddy-Barraud could you please rename the classes to be consistent with the source files, `PairDPDCharged` to `PairDPDCoulSlaterLong` and their GPU counterpart? Similarly, the class in the GPU library would be `DPDCoulSlaterLong`, instead of `DPDCharged`. Thanks!

No problem, you already said it by email, and I totally understand. It's done and working.

**Comment by ndtrung81** [comment]
@Eddy-Barraud looks good to me.

**Review by ndtrung81 — APPROVED** [review]

**Comment by akohlmey** [comment]
@Eddy-Barraud I will need to have a closer look at your pull request and make some changes. It looks like there is a problem with reading binary restarts. I also would like to change the order of arguments to the pair style to have the cutoffs come last and make the coulomb cutoff optional with the default the DPD cutoff (like for other styles like lj/cut/coul/long).

File: src/DPD-BASIC/pair_dpd_coul_slater_long.cpp [hunk 3]
```diff
@@ -0,0 +1,573 @@
...
+  for (i = 1; i <= atom->ntypes; i++)
+    for (j = i; j <= atom->ntypes; j++) {
+      if (me == 0) utils::sfread(FLERR,&setflag[i][j],sizeof(int),1,fp,nullptr,error);
+      MPI_Bcast(&setflag[i][j],1,MPI_INT,0,world);
+      if (setflag[i][j]) {
+        if (me == 0) {
+          utils::sfread(FLERR,&a0[i][j],sizeof(double),1,fp,nullptr,error);
+          utils::sfread(FLERR,&gamma[i][j],sizeof(double),1,fp,nullptr,error);
+          utils::sfread(FLERR,&cut[i][j],sizeof(double),1,fp,nullptr,error);
+          utils::sfread(FLERR, &scale[i][j],sizeof(double),1,fp, nullptr, error);
```
Comment (ndtrung81) [inline src/DPD-BASIC/pair_dpd_coul_slater_long.cpp]: Axel is correct, proc 0 needs to read in `cut_dpd[i][j]`, `cut_dpdsq[i][j]`, `cut_slater[i][j]` and `cut_slatersq[i][j]` here

**Comment by Eddy-Barraud** [comment]
@ndtrung81 @akohlmey I forgot to try restarting a simulation. I understand the logic of optional coulomb cutoff. I stay available if you want me to change something @akohlmey

**Pull request merged**